doc/source/programs/gdalwarp.rst

```diff
@@ -493,6 +502,164 @@ The details of how it is taken into account depends on the resampling kernel:
+increase will be significant. Disks are the slowest part of the process.
+By default :program:`gdalwarp` won't take much advantage of RAM. Using the flag
+:option:`-wm 500` will operate on 500MB chunks at a time which is better than the
+default. Increasing the I/O block cache size may also help. This can be done by
+setting the :config:`GDAL_CACHEMAX` configuration like:
+
+.. code-block:: bash
+
+   gdalwarp --config GDAL_CACHEMAX 500 -wm 500 ...
+
+This uses 500MB of RAM for read/write caching, and 500MB of RAM for working
+buffers during the warp. Beyond that it is doubtful more memory will make a
+substantial difference.
+
+Check CPU usage while :program:`gdalwarp` is running. If it is substantially
+less than 100% then you know things are IO bound. Otherwise they are CPU bound.
+The ``--debug`` option may also provide useful information. For instance, after
+running the following:
+
+.. code-block::
+
+   gdalwarp --debug on abc.tif def.tif
+
+a message like the following will be output:
+
+::
+
+  GDAL: 224 block reads on 32 block band 1 of utm.tif
+
+In this case it is saying that band 1 of :file:`utm.tif` has 32 blocks, but
+that 224 block reads were done, implying that lots of data was having to be
+re-read, presumably because of a limited IO cache. You will also see messages
+like:
+
+::
+
+   GDAL: GDALWarpKernel()::GWKNearestNoMasksByte()
+   Src=0,0,512x512 Dst=0,0,512x512
+
+The Src/Dst windows show you the "chunk size" being used. In this case my whole
+image which is very small. If you find things are being broken into a lot of
+chunks increasing :option:`-wm` may help somewhat.
+
+But far more important than memory are ensuring you are going through an
+optimized path in the warper. If you ever see it reporting
+``GDALWarpKernel()::GWKGeneralCase()`` you know things will be relatively slow.
+Basically, the fastest situations are nearest neighbour resampling on 8bit data
+without nodata or alpha masking in effect.
+
+.. caution::
+
+   Increasing the value of the :option:`-wm` param may lead to loss of
+   performance in certain circumstances, particularly when gdalwarping *many*
+   small rasters into a big mosaic. See http://trac.osgeo.org/gdal/ticket/3120
+   for more details
+
+In some cases using ``--config GDAL_CACHEMAX xxx -wm xxx`` can be detrimental.
+A symptom of this is:
+
+::
+
+   ERROR 2: Out of memory allocating 365425784 byte destination buffer.
+
+Our understanding is that 32 bit processes are frequently subject to memory
+fragmentation problems and so even in a process with - in theory - 2GB of heap
+RAM space available it is still often difficult to allocate large blocks of
+memory. In this case with :option:`-wm 500` set the main warp buffers can be
+quite large (364MB for the example) and it seems it has failed to allocate the
+buffer. Use more modest buffers (or not use the options at all) or else get a
+64 bit executable for :program:`gdalwarp`.
```

Comment:
Maybe this discussion about 32-bit systems can be removed?

Reply:
yes, that can be removed (still true for the folks on 32bit platforms, but irrelevant to most)